### PR TITLE
Add support for recoverying from gaps

### LINKED ISSUE #1136: Relax handling of partial connections in spicy-plugin
Now that a first iteration of error recovery has [landed in in Spicy](https://github.com/zeek/spicy/pull/1054) we should revisit how spicy-plugin handles partial connections/undelivered data. It might even be possible to implement the first pieces of this before zeek/spicy#1135 has been addressed.

The old prototype implemented this by injecting a special stream chunk for missing data. Once the parser hits that chunk it could automatically enter search mode (i.e., w/o waiting to first hit the parser error).

Implementation plan for zeroth iteration:
- teach gaps to `Stream`
- if parser hits gap trigger parse error
- spicy-plugin passes on data with undelivered data/partial connections
- trigger usual sync handling if unit supports it, or propagate up

### PR DESCRIPTION
**Pull request by bbannier**
Closes #1136.

### CODE REVIEW

File: hilti/runtime/include/types/stream.h [hunk 1]
```diff
@@ -85,8 +90,7 @@ class Chunk {
     using Array = std::pair<Size, std::array<Byte, SmallBufferSize>>;
     using Vector = std::vector<Byte>;
 
-    Chunk(Offset o, std::array<Byte, SmallBufferSize>&& d, Size n)
-        : _offset(o), _data(std::make_pair(n, std::move(d))) {}
+    Chunk(Offset o, std::array<Byte, SmallBufferSize> d, Size n) : _offset(o), _data(std::make_pair(n, d)) {}
```
Comment (rsmmr) [inline hilti/runtime/include/types/stream.h:93]: Why this change?

File: hilti/runtime/include/types/stream.h [hunk 2]
```diff
@@ -105,24 +112,28 @@ class Chunk {
         _offset = other._offset;
         _data = std::move(other._data);
         _next = std::move(other._next);
-        _chain = std::move(other._chain);
+        _chain = other._chain;
```
Comment (rsmmr) [inline hilti/runtime/include/types/stream.h:115]: Similar question: what's the reason for this change?

File: hilti/runtime/include/types/stream.h [hunk 3]
```diff
@@ -105,24 +112,28 @@ class Chunk {
...
         return *this;
     }
 
     ~Chunk() = default;
 
     Offset offset() const { return _offset; }
     Offset endOffset() const { return _offset + size(); }
-    bool isCompact() const { return std::holds_alternative<Array>(_data); }
+    bool isCompact() const { return ! std::holds_alternative<Vector>(_data); }
```
Comment (rsmmr) [inline hilti/runtime/include/types/stream.h:123]: According to a quick grep, it doesn't look like we're using this method anywhere. Given that the semantics are getting a bit odd now, I'd just remove it.

File: hilti/runtime/include/types/stream.h [hunk 4]
```diff
@@ -220,16 +237,16 @@ class Chunk {
         if ( n <= Chunk::SmallBufferSize ) {
             std::array<Byte, Chunk::SmallBufferSize> x{};
             std::copy(ud, ud + n.Ref(), x.data());
-            return Chunk(o, std::move(x), n);
+            return Chunk(o, x, n);
```
Comment (rsmmr) [inline hilti/runtime/include/types/stream.h:240]: Once more: wondering about this change?

File: hilti/runtime/include/types/stream.h [hunk 5]
```diff
@@ -1410,9 +1434,9 @@ class Stream {
      */
     explicit Stream(const char* d) : Stream(Chunk(0, d, strlen(d))) {}
 
-    /** Creates an instance from an existing memory block. The data will be copied. */
-    Stream(const char* d, Size n) : Stream(Chunk(0, d, n)) {}
-
+    /** Creates an instance from an existing memory block. The data will be copied if set, otherwise a gap will be
+     * recorded. */
```
Comment (rsmmr) [inline hilti/runtime/include/types/stream.h:1443]: (Fix comment formatting.)

File: hilti/runtime/src/types/stream.cc [hunk 6]
```diff
@@ -173,6 +173,27 @@ int Chain::numberOfChunks() const {
...
+        if ( ! c )
+            break;
+
+        // Try again at the offset of the next chunk.
+        o = c->offset();
+    }
+
+    // If we haven't found a non-gap chunk simply return a view at the next
+    // byte. Since this is a gap chunk this can cause recovery in the caller.
+    return advance(1U);
```
Comment (rsmmr) [inline hilti/runtime/src/types/stream.cc:194]: I find this logic hard to follow because it's mixing operation on chunks with offset-based computation. I believe it's also not the most efficient. Would it be easier to create an iterator `i = unsafeBegin(); ++i` at the beginning, then work with the chunk `c` that that's pointing to & update `i` accordingly each round; and eventually return a new view `(i, end())`?

**Review by rsmmr — COMMENTED** [review]
For the 1st commit, I ended up wondering about a couple of move-semantics changes; no need to change anything there if those are all as intended, just checking.

File: hilti/runtime/src/types/stream.cc (hunk 6 above)
Comment (bbannier) [reply hilti/runtime/src/types/stream.cc:194]: Simplified liked discussed out of band.

File: hilti/runtime/include/types/stream.h (hunk 2 above)
Comment (bbannier) [reply hilti/runtime/include/types/stream.h:115]: This fixes a clang-tidy warning about an unnecessary `move` since the array here is trivially copyable. I'll move this to its own patch.

File: hilti/runtime/include/types/stream.h (hunk 4 above)
Comment (bbannier) [reply hilti/runtime/include/types/stream.h:240]: This is related to the signature change in the constructor above which cleans up a clang-tidy warning. I'll move this to its own patch.

File: hilti/runtime/include/types/stream.h (hunk 1 above)
Comment (bbannier) [reply hilti/runtime/include/types/stream.h:93]: This fixes a clang-tidy warning about an unnecessary `move` since the array here is trivially copyable. I'll move this to its own patch.

File: hilti/runtime/include/context.h [hunk 7]
```diff
@@ -85,20 +85,13 @@ extern Context*& current();
 extern Context* master();
 
 /** Returns the context set for the current hardware thread. */
-inline auto get() {
-    assert(current());
```
Comment (rsmmr) [inline hilti/runtime/include/context.h:89]: Suggestion: Rather than completely disabling the null check, how about adding a boolean parameter `allow_missing_context` with default `false`. We'd then set that to true from just those two locations where it can happen for benign reasons.

File: hilti/runtime/include/types/stream.h [hunk 8]
```diff
@@ -1389,7 +1389,14 @@ class View {
...
 inline std::ostream& operator<<(std::ostream& out, const View& x) {
-    out << escapeBytes(x.data().str());
+    Bytes data;
+    try {
+        data = x.data();
+    } catch ( hilti::rt::MissingData& ) {
+        data.append("<missing data>");
+    }
+
+    out << escapeBytes(data.str());
```
Comment (rsmmr) [inline hilti/runtime/include/types/stream.h:1399]: As discussed, we should change this to insert `<gap>` (or so) into the rendered data, instead of fully replacing it.

**Review by rsmmr — DISMISSED** [review]

**Comment by rsmmr** [comment]
"Approve" is for "changes so far", let's still to the review of if we're handling exceptions correctly internally,  potentially with extending/changing the internal API for better control.

**Review by rsmmr — DISMISSED** [review]
Looks good on the final changes. Only thing is that it'd be good to add a test for `dataForPrint()`, didn't see that.
